Add ndm build base image Dockerfile and github workflows

z0marlin opened the pull request:
- Add Dockerfile for base image to be used to build ndm binaries. The image contains all deps required to build ndm.
- Add github workflows to build the base image on PRs and releases.

Fixes openebs/node-disk-manager#628

Signed-off-by: Aditya Jain <aditya.jainadityajain.jain@gmail.com>

z0marlin committed:
Add ndm build base image Dockerfile and github workflows

Signed-off-by: Aditya Jain <aditya.jainadityajain.jain@gmail.com>

z0marlin commented on .github/workflows/pull-request.yaml:
```diff
@@ -0,0 +1,47 @@
+name: ci
```
@akhilerm which license header should I add in these files? The same one used in ndm?

akhilerm replied:
same as in NDM.

z0marlin commented on .github/workflows/pull-request.yaml:
```diff
@@ -0,0 +1,47 @@
+          file: ./build/ndm-build-base/Dockerfile
+          push: false
+          load: true
+          platforms: linux/amd64
```
need to update the platforms list here

akhilerm replied:
please update this with arm and ppc platforms. All platforms for which we build NDM should be added here.

akhilerm commented on .github/workflows/pull-request.yaml:
```diff
@@ -0,0 +1,47 @@
+      - name: Run Trivy vulnerability scanner
+        uses: aquase--recursivecurity/trivy-action@master
+        with:
+          image-ref: 'openebs/ndm-build-base:ci'
```
How will this work? ci image is not the currently pushed image.

z0marlin replied:
The image is built and tagged in the same job right. This is a step in the same job, so it will run on the same runner, correct?

akhilerm replied:
Yes. Correct.

akhilerm commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,116 @@
+name: release
```
Do we need to do a release for NDM base, shall we base it on commit IDs

z0marlin replied:
we can change it to every commit that is pushed to the main branch, but that will result in too many images (one for each commit). Is that acceptable?

akhilerm replied:
The dependencies dont change much. Also, we need to have an image whenever a dependency is updated, since that image needs to be used in the NDM repository.

akhilerm commented on build/ndm-build-base/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+FROM debian:buster as build
+RUN apt-get update && apt-get install -y make git gcc
+WORKDIR /tmp/openSeachest
```
we should clone from seachest org and use it instead of using local copy

z0marlin replied:
The `install-dep` script in ndm [uses the openebs repository](https://github.com/openebs/node-disk-manager/blob/9fe3968089a560cb1dde0269401778e86c7fd772/build/install-dep.sh#L29), so I used the same. Won't it be good to use this repo itself? This way we can control what code we pull in from seachest to build ndm.

akhilerm replied:
We need to checkout to the specific version of openseachest . Where is it being done?

z0marlin replied:
ideally, this should be done by using the image which was tagged at the specific version right? I mean we shouldn't be specifying any version here. The consumer should pull in the image corresponding to the commit, right? Right now ndm is pulling `Release-19.06.02`. We might want to trigger a pipeline run for this commit?

akhilerm replied:
But we also have other dependencies.  How will that be handled? For Release-19.06.02, we have with base image ubuntu 18 and ubuntu 20, how will this be handled?

akhilerm reviewed (changes requested):
given some comments

z0marlin committed:
Add license header to new files

Signed-off-by: Aditya Jain <aditya.jainadityajain.jain@gmail.com>

akhilerm commented on .github/workflows/pull-request.yaml:
```diff
@@ -0,0 +1,61 @@
+        with:
+          file: ./build/ndm-build-base/Dockerfile
+          push: false
+          load: true
```
Since load is set as true, only amd64 will be loaded after building . rest all images will be discarded.

akhilerm commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,130 @@
+        with:
+          version: v0.5.1
+
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      - name: Login to Quay
+        uses: docker/login-action@v1
+        with:
+          registry: quay.io
+          username: ${{ secrets.QUAY_USERNAME }}
+          password: ${{ secrets.QUAY_TOKEN }}
+
+      - name: Login to GHCR
```
Is it needed to have the base image in all registries? Since the base image is something that we mention in the script , it will always be pulled from docker hub right?

z0marlin replied:
ok, let's only keep it in ghcr?

akhilerm reviewed (commented):
@z0marlin Added a few more comments.

z0marlin committed:
Push ndm build base image to ghcr only

Signed-off-by: Aditya Jain <aditya.jainadityajain.jain@gmail.com>

z0marlin committed:
Release new image for every new commit

Build and push new ndm build base image for every new commit pushed to
the develop branch or the release tags matching the expression -
"Release-**"

Signed-off-by: Aditya Jain <aditya.jainadityajain.jain@gmail.com>

z0marlin committed:
Build ndm base image on all supported platforms on PR

Signed-off-by: Aditya Jain <aditya.jainadityajain.jain@gmail.com>

z0marlin committed:
Lock openSeaChest version used in ndm base image

Pull specific version of openSeaChest in the Dockerfile. Run workflows
only if there are changes in `build/ndm-build-base/**`.

Signed-off-by: Aditya Jain <aditya.jainadityajain.jain@gmail.com>

akhilerm commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,115 @@
+            ghcr.io/${{ env.IMAGE_ORG }}/ndm-build-base
+          tag-latest: true
+          tag-semver: |
+            {{version}}
```
What will the tags that we get in this case, This is based on the semantic version tags that are present right?  Dont we need the commit-hash / date something like that as the image tags?

akhilerm reviewed (commented):
looks good. A final question on the tags to be used for the base images.

z0marlin committed:
Fix ndm base image tags when building new image

Signed-off-by: Aditya Jain <aditya.jainadityajain.jain@gmail.com>

z0marlin committed:
Rename release.yml to build.yml

Signed-off-by: Aditya Jain <aditya.jainadityajain.jain@gmail.com>

z0marlin committed:
Add build date as a tag to the ndm build base image

Signed-off-by: Aditya Jain <aditya.jainadityajain.jain@gmail.com>

akhilerm commented on .github/workflows/build.yml:
```diff
@@ -39,7 +39,8 @@ jobs:
         id: date
         run: |
-          echo "::set-output name=DATE::$(date -u +'%Y-%m-%dT%H:%M:%S%Z')"
+          echo "::set-output name=timestamp::$(date -u +'%Y-%m-%dT%H:%M:%S%Z')"
+          echo "::set-output name=date::$(date -u +'%Y%m%d')"
```
use upper case for environment variables.

akhilerm replied:
```suggestion
          echo "::set-output name=DATE::$(date -u +'%Y-%m-%dT%H:%M:%S%Z')"
```

z0marlin committed:
Use date global provided by meta action for setting date tag

akhilerm commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,113 @@
+      - name: Set Build Date
+        id: date
+        run: |
+          echo "::set-output name=timestamp::$(date -u +'%Y-%m-%dT%H:%M:%S%Z')"
```
```suggestion
          echo "::set-output name=DATE::$(date -u +'%Y-%m-%dT%H:%M:%S%Z')"
```

akhilerm commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,113 @@
+          tags: |
+            ${{ steps.docker_meta.outputs.tags }}
+          build-args: |
+            DBUILD_DATE=${{ steps.date.outputs.timestamp }}
```
```suggestion
            DBUILD_DATE=${{ steps.date.outputs.DATE }}
```

akhilerm committed:
Update .github/workflows/build.yml

akhilerm committed:
Update .github/workflows/build.yml

akhilerm reviewed (approved):
LGTM
